src/pynwb/__init__.py

```diff
@@ -139,24 +159,100 @@ def load_namespaces(**kwargs):
     return __TYPE_MAP.load_namespaces(namespace_path)
 
-
-# load the core namespace, i.e. base NWB specification
-__resources = __get_resources()
-if os.path.exists(__resources['namespace_path']):
-    load_namespaces(__resources['namespace_path'])
-else:
-    raise RuntimeError(
-        "'core' is not a registered namespace. If you installed PyNWB locally using a git clone, you need to "
-        "use the --recurse_submodules flag when cloning. See developer installation instructions here: "
-        "https://pynwb.readthedocs.io/en/stable/install_developers.html#install-from-git-repository"
-    )
-
-
 def available_namespaces():
     """Returns all namespaces registered in the namespace catalog"""
     return __NS_CATALOG.namespaces
 
 
+def _git_cmd(*args) -> subprocess.CompletedProcess:
+    """
+    Call git with the package as the directory regardless of cwd.
+
+    Since any folder within a git repo works, don't try to ascend to the top, since
+    if we're *not* actually in a git repo we're only guaranteed to know about
+    the inner `pynwb` directory.
+    """
+    parent_dir = str(Path(__file__).parent)
+    result = subprocess.run(["git", "-C", parent_dir, *args], capture_output=True)
+    return result
+
+
+def _clone_submodules():
+    if _git_cmd('rev-parse').returncode == 0:
+        warnings.warn(
+            'NWB core schema not found in cloned installation, initializing submodules...',
+            stacklevel=1)
+        res = _git_cmd('submodule', 'update', '--init', '--recursive')
+        if not res.returncode == 0:  # pragma: no cover
+            raise RuntimeError(
+                'Exception while initializing submodules, got:\n'
+                'stdout:\n' + ('-'*20) + res.stdout + "\nstderr:\n" + ('-'*20) + res.stderr)
+    else:  # pragma: no cover
+        raise RuntimeError("Package is not installed from a git repository, can't clone submodules")
+
+
+def _load_core_namespace(final:bool=False):
+    """
+    Load the core namespace into __TYPE_MAP,
+    either by loading a pickled version or creating one anew and pickling it.
+
+    We keep a dotfile next to it that tracks what version of pynwb created it,
+    so that we invalidate it when the code changes.
+
+    Args:
+        final (bool): This function tries again if the submodules aren't cloned,
+            but it shouldn't go into an infinite loop.
+            If final is ``True``, don't recurse.
+    """
+    global __NS_CATALOG
+    global __TYPE_MAP
+    global __resources
+
+    # if we have a version indicator file and it doesn't match the current version,
+    # scrap the cached typemap
+    if os.path.exists(__resources['cached_version_indicator']):
+        with open(__resources['cached_version_indicator'], 'r') as f:
+            cached_version = f.read().strip()
+        if cached_version != __version__:
+            Path(__resources['cached_typemap_path']).unlink(missing_ok=True)
+
+    # load pickled typemap if we have one
```

Comment:
I'm a little wary of loading the pickled typemap if the cached_version_indicator file doesn't exist for some reason. I would rather play it safe and re-load the type map in this case. Can you adjust the logic for that?

Reply:
got it in [`a34ef20`](https://github.com/NeurodataWithoutBorders/pynwb/pull/1931/commits/a34ef20c5b01a258c2bfe339ce01539267def2bc)